Step 1 optimizations

iegorval opened the pull request:
# Pull request - step 1 optimization

Hello. I came to notice that there are a several things that might be better done in data loading process. I am not a big C++ expert, so I recommend a double-check and be available also to discuss these changes in more details. If you tell me on which datasets I should test, I can do that also.

1) The main thing I checked was the time filtering of the point cloud to split it to different worker data. I pre-filter the points by timestamp and then use logarithmic lower/upper bound functions to locate the range of points to use for filtering. The resulting time measurements (of course, very HW and system-dependent but clearly having order of magnitude difference) are below. Note that sorting does not seem to have a major overhead compared to the gained speed at filtering. Faster search algorithm (if timestamps are represented as integer nanoseconds) can be considered. 

* Elapsed time reading: 130874ms (original code)
* Elapsed time reading: 154925ms (with sorting)

* Elapsed time timestamp filtering: 172041ms (original code)
* Elapsed time timestamp filtering: 158ms (applied on sorted points)

I have seen that you have some pretty similar code in the code you commented out. Have you found some situation where proposed solution does not work, or why was it abandoned? 

2) I also commented out writing of `calibrationValidation` file in `lidar_odometry_gui` optional. That makes a huge difference reducing time of reading (lambda function) from 172041ms to 48553ms (including sorting).  The overhead might be small for small files, but once one big laz is used, the overhead is just huge. What if we added some parameter to the interface telling whether to save the calibration validation or not? (I can add it.)

3) There seem to be a bit of a mess with creating `WorkerData`, which always have length of attributes depending on the fixed `threshold` parameter. A lot of memory is constantly cleared and reserved manually for `WorkerData` while this can be dealt with by just changing a bit the whole for-loop and just keeping `WorkerData` inside its scope. The execution of the code dealing with filling of `WorkerData`'s vector then goes from 7307ms to 6265ms. The change is small but consistent between different runs, at least on my computer.

I also noticed that you separate data into different `WorkerData` as if to pass it to multiprocessing, but then just assemble it back to a long vector in e.g. step 2. Is the multiprocessing abandoned? Some issues? No time to implement? Because otherwise I am not sure why to do split-and-merge. It is concerning me as very often I get BSOD at further steps, especially at Consistency. I would happy to check if something can be done there in my free time, if you would be interested.

Best regards,
Valeria

JanuszBedkowski commented on apps/lidar_odometry_step_1/lidar_odometry.cpp:
```diff
@@ -324,16 +324,16 @@ void lidar_odometry_gui()
-                                       if (fn == laz_files.front())
-                                       {
-                                           fs::path calibrationValidtationFile = wdp / "calibrationValidation.asc";
+                                    //    if (fn == laz_files.front())
```
Please do not remove this functionality. It is crutial for calibration validation. Try to consider saving e.g. 1000.000 points.

JanuszBedkowski commented on apps/lidar_odometry_step_1/lidar_odometry.cpp:
```diff
@@ -324,16 +324,16 @@ void lidar_odometry_gui()
+                                    //        fs::path calibrationValidtationFile = wdp / "calibrationValidation.asc";
 
-                                           std::ofstream testPointcloud{calibrationValidtationFile.c_str()};
-                                           for (const auto &p : data)
```
Here You can save first 1000.000 ponts

michalpelka replied:
We can save it only when multi-lidar setup is loaded.

JanuszBedkowski commented:
Lidar odometry is crashing on my machine. Please check it on windows.

iegorval commented:
> Lidar odometry is crashing on my machine. Please check it on windows.

Hello @JanuszBedkowski,

Thanks for commenting on the pull request.

This was caused by a small bug introduced by me while making the final commit indeed. Realized there was a line of code missing.

But what I also noticed while debugging is this original for cycle you have for filtering:
![image](https://github.com/user-attachments/assets/b5e0f818-a458-4817-b177-7498f56ade6e)
Is it indeed meant to be like this, with that `break` on line 518? If my intuition is correct, you have this `break` just to make the search/filtering stop earlier. But actually this `break` changes the result (on my data at least): more points are filtered out with this `break` that without it, which from condition in line 507 seems like not something that I would expect.

Could you please comment on this one before I make the commit?

iegorval commented:
@JanuszBedkowski @michalpelka 

So, I added some changes:

1. I fixed the crash (we tested also on Windows machine). The crash was connected with me accidentally removing the `std::sort` which I still mentioned in the text of the pull request. Incidentally it also showed that the assumption of having all points perfectly sorted by timestamp is incorrect. I also wrote some Python code (as I am better-versed in Python tbh) to check the timestamp filtering and I found out that we throw out ~1% points by assuming that timestamps are aligned (in the original version of the timestamp filtering). As my code does not have this assumption and force-sorts the data, it regains these points. While this is not a huge amount of points, I also do not see reason of getting rid of them. Unless it was by some design choice I am not aware of.
2. While debugging my own problem and cross-checking with Python, I also noticed that there is always one more IMU data row than it has to be. It seems that the file reader picks the empty line on IMU csv file, and while the line is empty, still uses previously-read data to add to the IMU data list. So, while using e.g. IMU 0, we always have at the very end of IMU data, one single row from IMU 1. Again, probably it does not have some big influence, considering the total amount of the IMU poses, but that was probably not done on purpose.

JanuszBedkowski commented:
Good job THX!
